python/cudf/cudf/core/index.py

```diff
@@ -2661,6 +2664,101 @@ def categories(self):
         return self._values.cat().categories
 
 
+class interval_range(GenericIndex):
+    """
+    Returns a fixed frequency IntervalIndex.
+
+    Parameters
+    start : numeric, default None
+    Left bound for generating intervals.
+
+    end : numeric , default None
+    Right bound for generating intervals.
+
+    freq : numeric, default None
+    The length of each interval. Must be consistent
+    with the type of start and end
+
+    name : str, default None
+    Name of the resulting IntervalIndex.
+
+    closed : {‘left’, ‘right’, ‘both’, ‘neither’}, default ‘right’
+    Whether the intervals are closed on the left-side, right-side,
+    both or neither.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, start=None, end=None, freq=1, closed="right", name=None,
+    ) -> "interval_range":
+        left_col = cupy.arange(start, end, freq)
+        right_col = cupy.arange(start + freq, end + 1, freq)
+        # sometimes the left col overlaps with the right col,
+        # seems to occur with even freq and odd end, and in other places too
+        if len(left_col) != len(right_col):
+            left_col = cupy.arange(start, end - freq, freq)
+        if len(right_col) == 0 or len(left_col) == 0:
+            # I get an error passing to cudf IntervalIndex
+            return pd.IntervalIndex([], closed=closed)
```

Comment:
I'm not sure what to do in this case because when IntervalIndex is passed an empty list we expect it to produce an empty index like pandas. Instead, I'm getting an error that seems to be saying `struct` can't take any inputs that are null. You can find an example of the error below. Let me know if there's anything I should change or add here. Thank you! 

```
In [1]: import cudf

In [2]: import pandas as pd

In [3]: pd.IntervalIndex([])
Out[3]: 
IntervalIndex([],
              closed='right',
              dtype='interval[int64]')

In [4]: cudf.IntervalIndex([])
Out[4]: ---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/compose/etc/conda/cuda_10.2/envs/rapids/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/compose/etc/conda/cuda_10.2/envs/rapids/lib/python3.7/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395 
    396             return _default_pprint(obj, self, cycle)

~/compose/etc/conda/cuda_10.2/envs/rapids/lib/python3.7/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/cudf/python/cudf/cudf/core/index.py in __repr__(self)
   2000                 output = output.replace("'", "")
   2001         else:
-> 2002             output = preprocess.to_pandas().__repr__()
   2003 
   2004         # Fix and correct the class name of the output

~/cudf/python/cudf/cudf/core/index.py in to_pandas(self)
    659         <class 'cudf.core.index.GenericIndex'>
    660         """
--> 661         return pd.Index(self._values.to_pandas(), name=self.name)
    662 
    663     def tolist(self):

~/cudf/python/cudf/cudf/core/column/column.py in to_pandas(self, index, nullable, **kwargs)
    138         elif is_interval_dtype(self.dtype):
    139             pd_series = pd.Series(
--> 140                 pd.IntervalDtype().__from_arrow__(self.to_arrow())
    141             )
    142         else:

~/cudf/python/cudf/cudf/core/column/interval.py in to_arrow(self)
     53     def to_arrow(self):
     54         typ = self.dtype.to_arrow()
---> 55         return pa.ExtensionArray.from_storage(typ, super().to_arrow())
     56 
     57     def copy(self, deep=True):

~/compose/etc/conda/cuda_10.2/envs/rapids/lib/python3.7/site-packages/pyarrow/array.pxi in pyarrow.lib.ExtensionArray.from_storage()

TypeError: Incompatible storage type struct<left: null, right: null> for extension type extension<pandas.interval>
```

Reply:
Hi @brandon-b-miller, this is the issue I mentioned. Reading it again, it looks like its an issue of struct and pandas interval not working well together. Anyway, let me know if you have any ideas about how to approach fixing this :D

Reply:
I realised this morning as I was reading more on `structs` that they are not the same as `scalars` 😅 Sorry, Brandon, I'm remembering you worked on scalars and Ashwin worked on structs. For some reason, in the meetingI was mixing the two up! It probably makes sense to check with Ashwin about these questions😄